doc/experimental_ui_api.md

```diff
@@ -0,0 +1,2027 @@
+    return ui.Window "gui" {
+        root = ui.Root {
+            size = {108, 40},
+            padding = {4},
+
+            box_fill = "black#8C",
+
+            ui.Button "counter" {
+                box_fill = "maroon",
+                label = "Clicks: " .. state.num_clicks,
+
+                on_press = function(ev)
+                    state.num_clicks = state.num_clicks + 1
+                    context:update()
+                end,
+            },
+        },
+    }
+end
+
+core.register_on_joinplayer(function(player)
+    local state = {num_clicks = 0}
+    ui.Context(builder, player:get_player_name(), state):open()
+end)
+```
+
+The state table is primarily for persistent data. However, it is often useful
+to send temporary data that only applies to a single call of the builder
+function, such as when setting persistent fields. This can be provided in the
+form of a _parameter table_. Parameter tables can be provided to the methods
+`open()`, `update()`, and `reopen()` and will be passed directly to the builder
+function. After the builder function returns, the parameter table is discarded.
+
+The following is a basic example that sets a persistent field when first
+opening the window by using parameter tables (note that it uses a slider, which
+is not implemented yet):
+
+```lua
+local function builder(context, player, state, param)
+    return ui.Window "gui" {
+        root = ui.Root {
+            size = {500, 40},
+            padding = {4},
+
+            box_fill = "black#8C",
+
+            ui.Slider "percent" {
+                label = state.scroll .. "%",
+
+                min = 0,
+                max = 100,
+
+                -- `param.set_scroll` is only set in `open()`, not in the
+                -- `update()` in `on_scroll`, so `value` is only set to
+                -- `state.scroll` when the window is first opened.
+                value = param.set_scroll and state.scroll,
+
+                on_scroll = function(ev)
+                    state.scroll = ev.value
+                    context:update()
+                end,
+            },
+        },
+    }
+end
+
+core.register_on_joinplayer(function(player)
+    local state = {scroll = 50}
+    local param = {set_scroll = true}
+
+    ui.Context(builder, player:get_player_name(), state):open(param)
+end)
+```
+
+Utility functions
+-----------------
+
+* `ui.new_id()`: Returns a new unique ID string.
+    * **This function may not be used for elements that require an ID to be
+      provided, such as buttons or edit field elements!** These elements
+      require an ID that will stay constant across window updates!
+    * It is usually not necessary to use this function directly since the API
+      will automatically generate IDs when none is required.
+    * The format of this ID is not precisely specified, but it will have the
+      format of an engine reserved ID and will not conflict with any other IDs
+      generated during this session.
```

Comment:
ui.new_id() gives a number of 0->large. What If I also give an id that is just a number "1"?

Reply:
No, it gives an ID of `_0`, which is an engine-reserved ID. Mods are not allowed to use IDs with leading underscores or dashes.